src/main/java/org/opensearch/ad/ratelimit/CheckpointReadWorker.java

```diff
@@ -386,31 +386,35 @@ private ActionListener<Optional<AnomalyDetector>> onGetDetector(
             }
 
             if (result != null && result.getRcfScore() > 0) {
-                AnomalyResult resultToSave = result
-                    .toAnomalyResult(
+                RequestPriority requestPriority = result.getGrade() > 0 ? RequestPriority.HIGH : RequestPriority.MEDIUM;
+
+                List<AnomalyResult> resultsToSave = result
+                    .toIndexableResults(
                         detector,
                         Instant.ofEpochMilli(origRequest.getDataStartTimeMillis()),
                         Instant.ofEpochMilli(origRequest.getDataStartTimeMillis() + detector.getIntervalInMilliseconds()),
                         Instant.now(),
                         Instant.now(),
                         ParseUtils.getFeatureData(origRequest.getCurrentFeature(), detector),
-                        entity,
+                        Optional.ofNullable(entity),
                         indexUtil.getSchemaVersion(ADIndex.RESULT),
                         modelId,
                         null,
                         null
                     );
 
-                resultWriteQueue
-                    .put(
-                        new ResultWriteRequest(
-                            origRequest.getExpirationEpochMs(),
-                            detectorId,
-                            result.getGrade() > 0 ? RequestPriority.HIGH : RequestPriority.MEDIUM,
-                            resultToSave,
-                            detector.getCustomResultIndex()
-                        )
-                    );
+                for (AnomalyResult r : resultsToSave) {
+                    resultWriteQueue
```

Comment:
Looks good - one q: will there be conflicts in the write queues if there is both forecasting and AD (say sometime down the road, for the same detector). Forecasting can have a lot of writes.

Reply:
There can be conflicts as all writes share the same queue in a node. Indeed, forecasting is not cheap.